### ISSUE TITLE
z>3 QSO templates often fail color cuts: doubly applying LyA absorption?

### ISSUE THREAD

**Issue by sbailey** [desihub/desisim#235]
First identified in #234: when we add Lyman-alpha absorption to QSO templates, they often fail the targeting color cuts for QSOs with 3.1 < z < 3.7 .  I suspect that we are double applying LyA mean absorption and thus are generating overly blue QSOs.  @xavier @andreufont @ngbusca could you take a look and confirm / fix?

**Comment by ngbusca** [comment]
@sbailey Could you put a snippet of code that illustrates the problem?

**Comment by andreufont** [comment]
I don't know if the PCA templates include mean absorption, if so we should
fix this.

But if this was the case, adding a double lya absorption would make the
quasars redder, isn't it? The forest eats blue light.

Andreu

**Comment by sbailey** [comment]
@andreufont yes, you are right.  Redder.

@ngbusca example of them failing:
```python
import desisim.templates
qsomaker = desisim.templates.QSO()
flux, wave, meta = qsomaker.make_templates(100, seed=1)
```

John's notebook has example spectra and g-r vs. r-z colors with and without LyAF at the bottom:

https://github.com/desihub/desisim/blob/master/doc/nb/QSO_templates.ipynb

I'm just guessing about the double applying LyAF mean absorption.  First task is to check whether the PCA templates have mean absorption or not (I'm guess they do).

**Comment by andreufont** [comment]
About that notebook: cell #8 it shows spectra for the 10 quasars, with and
without the forest. Blue-ward of Lyman alpha these are the same for all
quasars but the second, that has larger continuum after adding forest. I
could see this happening if we are rescaling the spectra after adding the
forest, to match the input magnitude, but then why does this only happens
for the second quasar?

Andreu

**Comment by ngbusca** [comment]
The plot below shows the median templates near z=2.5 and z=3.5. It looks like there is indeed absorption in the templates.

The way the templates are constructed is by sampling the pca coefficients of BOSS spectra near the input redshift. My guess is that those coeffs were obtained including the forest in the pca fits. If yes, then the fix would be to redetermine them masking the forest, which implies updating the pca input files.

![image](https://cloud.githubusercontent.com/assets/4727509/25151965/3516eb6c-2490-11e7-893e-a8f8fb37e269.png)

**Comment by ngbusca** [comment]
@andreufont by default, r-mags (5500 < lambda < 6500, approx) are used to normalize the spectra. The second spectrum is the only one that has some forest in the r-band.

**Comment by andreufont** [comment]
@ngbusca got it, thanks!

**Comment by ngbusca** [comment]
I think I found the problem. The way the lyman-alpha forest is introduced is put on top of the absorption below the lyman-limit, which, I think, doesn't make sense. Note that the current lyman-alpha absorption doesn't account for whole lyman series, which means that qsos will be artificially bluer (probably negligibly so).

I'm committing a tentative fix where I remove lyman-alpha absorption below the lyman-limit.

**Comment by andreufont** [comment]
@ngbusca, I'm not sure what you mean. In principle there is nothing wrong with having Lyman alpha forest down to the blue end of the spectrograph, right? Regardless of rest-frame wavelength.

**Comment by ngbusca** [comment]
The continua have the absorption beyond the lyman-limit already incorporated. We shouldn't add lyman-alpha absorption on top of that, I think...

**Comment by ngbusca** [comment]
Alternatively, we could add only lyman-alpha absorption, but it would get less and less realistic in the lyman-beta, lyman-gamma, ... It's trivial to turn off the lyman-limit absorption in the templates if that's the better solution.

**Comment by andreufont** [comment]
In order to address this, we need to understand better what the PCA-based templates really contain. 

We (LyaF working group) would like them to be what we would call quasar continua, i.e., the spectrum that we would get if the Universe was empty, but I'm not sure this is what we are getting.

Ignoring metal absorptions, we can describe the observed flux in a quasar spectrum as:

f(l) = C(l) F(l) = C(l) F_alpha(l) F_beta(l) ... F_limit(l)

where C(l) is the continuum at a given observed wavelength l, and the total transmitted flux fraction F(l) is a number between 0 and 1, that can be decomposed as the product of different transmitted flux fractions (the total optical depth is a sum of the different optical depths).

Not all pixels have contributions from all these terms. Pixels with a rest-frame wavelength above the Lya line have F=1 (ignoring metals), and pixels with a rest-frame wavelength between Lyb and Lya have F = F_alpha.  Only pixels with a rest-frame wavelength below the Lyman limit have all contributions. 

So if the PCA-templates don't explicitly take all these terms into account, and correct for these absorptions, what we will have are quasar spectrum that already have some mean absorption on top. 

But I don't see how the changes that you suggest are going to fix this.

**Comment by ngbusca** [comment]
@andreufont I think we need to be more pragmatic here, we need a solution soon for the data challenge.

The template maker has an explicit factor for the absorption at lambda < lyman-limit derived from measurments. The current master puts lyman-alpha on top of this absorption, I think this is not correct. With the code currently at our disposal we can do at least two things to fix that:

* remove the lyman-alpha absorption below the lyman-limit
* remove the lyman-limit absortpion below the lyman limit

Both options give bluer spectra between the lyman limit and lyman-alpha (since there only lyman alpha absorption). The second option gives also bluer spectra below the lyman-limit. So, I'm proposing to go for the first.

**Comment by ngbusca** [comment]
@sbailey : After some discussion offline with @andreufont we concluded that there's no double application of the lyman-alpha absorption in the region 912 A < lambda_rest < 1216 A.

In the region lambda_rest < 912 A there might be double application depending on the meaning of the "mean free path" factor applied below 912A. If this accounts for all absorption including non-ionizing interactions, then we have to modify the current implementation. If it's only ionizing interactions, then the current implementation should be ok. @profxj could you clarify what the mean free path factor accounts for?

**Comment by profxj** [comment]
@ngbusca you are correct.  The MFP only accounts for the continuum
Lyman limit opacity.  See http://adsabs.harvard.edu/abs/2009ApJ...705L.113P

**Comment by ngbusca** [comment]
Thanks @profxj 

So I guess it's correct to add the lyman-alpha absorption as it's currently done

**Comment by ngbusca** [comment]
I did found, however, that adding the lyman-alpha forest was cutting the spectra below 10,000 A, which is screwing up the z-color

![image](https://cloud.githubusercontent.com/assets/4727509/25202125/eb4fdd18-255c-11e7-90c3-e84a99d7223a.png)

I fixed this in the branch double_lya (and removed my previous wrong "fix"):

![image](https://cloud.githubusercontent.com/assets/4727509/25202202/2d110d62-255d-11e7-85d1-d6a086414ec5.png)

I can now produce spectra at z~3.5 without hitting the max iterations limit.

**Comment by andreufont** [comment]
@ngbusca Good catch! Given that we never need forest for lambda > 7000 or so, I didn't bother extending the forest skewers to lambda =10000 or so. I thought we would just use F=1 for these large wavelengths, but the interpolation function that we use doesn't do any extrapolation.

Your fix in the #236 looks good.

**Comment by moustakas** [comment]
Blarg, I can't believe I didn't spot-check the red end of the spectra.  Thanks for catching this, @ngbusca!

**Comment by moustakas** [comment]
Fixed in #236.
